src/main/java/com/epam/aidial/core/service/ResourceOperationService.java

```diff
@@ -0,0 +1,53 @@
+package com.epam.aidial.core.service;
+
+import com.epam.aidial.core.data.ResourceType;
+import com.epam.aidial.core.storage.BlobStorage;
+import com.epam.aidial.core.storage.ResourceDescription;
+import lombok.AllArgsConstructor;
+
+@AllArgsConstructor
+public class ResourceOperationService {
+
+    private final ResourceService resourceService;
+    private final BlobStorage storage;
+    private final InvitationService invitationService;
+    private final ShareService shareService;
+
+    public void moveResource(String bucket, String location, ResourceDescription source, ResourceDescription destination, boolean overwriteIfExists) {
+        if (source.isFolder() || destination.isFolder()) {
+            throw new IllegalArgumentException("Moving folders is not supported");
+        }
+
+        String sourceResourcePath = source.getAbsoluteFilePath();
+        String sourceResourceUrl = source.getUrl();
+        String destinationResourcePath = destination.getAbsoluteFilePath();
+        String destinationResourceUrl = destination.getUrl();
+
+        if (!storage.exists(sourceResourcePath)) {
```

Comment:
Why do you check storage for conversations/prompts? Use resource service instead. Also it is not atomic operation. Check ... waitttttttttttt.... write.